effectful/handlers/pyro.py

```diff
@@ -252,190 +265,264 @@ def __repr__(self):
+    _kwargs: dict
 
-        base_dims = dims[len(dims) - n_base :]
-        named_dims = dims[:n_named]
-        sample_dims = dims[n_named : n_named + n_sample]
+    def __init__(self, dist_constr: Type[TorchDistribution], *args, **kwargs):
+        self._args = (dist_constr,) + tuple(defterm(a) for a in args)
+        self._kwargs = kwargs
 
-        # shape: | named | sample_shape | batch_shape | event_shape |
-        # TODO: replace with something more efficient
-        pos_tensor = to_tensor(value, self.indices.keys())
+    @property
+    def op(self):
+        return call
 
-        # shape: | sample_shape | named | batch_shape | event_shape |
-        pos_tensor_r = torch.permute(pos_tensor, sample_dims + named_dims + base_dims)
+    @property
+    def args(self):
+        return self._args
 
-        return pos_tensor_r
+    @property
+    def kwargs(self):
+        return self._kwargs
 
-    def _from_positional(self, value: torch.Tensor) -> torch.Tensor:
-        # maximal value shape: | sample_shape | named | batch_shape | event_shape |
-        return self.naming.apply(value)
+    @property
+    def _base_dist(self):
+        return self.args[0](*self.args[1:])
 
     @property
     def has_rsample(self):
-        return self.base_dist.has_rsample
+        return self._base_dist.has_rsample
 
     @property
     def batch_shape(self):
-        return (
-            torch.Size([s for s in self.indices.values()]) + self.base_dist.batch_shape
-        )
+        return self._base_dist.batch_shape
 
     @property
     def event_shape(self):
-        return self.base_dist.event_shape
+        return self._base_dist.event_shape
 
     @property
     def has_enumerate_support(self):
-        return self.base_dist.has_enumerate_support
+        return self._base_dist.has_enumerate_support
 
     @property
     def arg_constraints(self):
-        return self.base_dist.arg_constraints
+        return self._base_dist.arg_constraints
 
     @property
     def support(self):
-        return self.base_dist.support
-
-    def __repr__(self):
-        return f"PositionalDistribution({self.base_dist})"
+        return self._base_dist.support
 
     def sample(self, sample_shape=torch.Size()):
-        return self._to_positional(self.base_dist.sample(sample_shape))
+        return self._base_dist.sample(sample_shape)
 
     def rsample(self, sample_shape=torch.Size()):
-        return self._to_positional(self.base_dist.rsample(sample_shape))
+        return self._base_dist.rsample(sample_shape)
 
     def log_prob(self, value):
-        return self._to_positional(
-            self.base_dist.log_prob(self._from_positional(value))
-        )
+        return self._base_dist.log_prob(value)
 
     def enumerate_support(self, expand=True):
-        return self._to_positional(self.base_dist.enumerate_support(expand))
+        return self._base_dist.enumerate_support(expand)
 
+    def __getitem__(self, key: Collection[Operation[[], int]]):
```

Comment:
I think this should be removed, it's not part of the base `Distribution` API (though I wish it was) and we don't want specific `Term` representations to leak into user code by departing from their semantic types' APIs. People can just use `named_distribution` directly.